### ISSUE TITLE
`Filter.watch` is awkward if you don't know how to use gevent (or even if you do)

### ISSUE THREAD

**Issue by pipermerriam** [pipermerriam/web3.py#106]
* Version: x.x.x
* Python: 2.7/3.4/3.5
* OS: osx/linux/win

### What was wrong?

Currently, this won't work.

```python
filter = web3.eth.filter('latest')
filter.watch(sys.stdout.write)

while filter.running:
    gevent.sleep(0)
```

The reason it doesn't work is that until the first `gevent.sleep(..)` call happens to yield the execution context, the filter's greenlet doesn't actually spin up.

### How can it be fixed.

I think that this could be fixed one or both of two ways.

* Within `Filter.watch` add a `gevent.sleep(0)` to allow the greenlet to spin up.  (This might be a bad idea.  It feels a little wrong but it might just work fine)
* Within the `Filter.watch` method set the `running` flag to `True` so that the pattern above will work.

#### Cute Animal Picture

![1356835726-0](https://cloud.githubusercontent.com/assets/824194/18921087/b3d1c4ba-8560-11e6-969e-6afaa73cf73b.jpg)

**Comment by rainbeam** [comment]
> Within Filter.watch add a gevent.sleep(0) to allow the greenlet to spin up. (This might be a bad idea. It feels a little wrong but it might just work fine)

In my limited gevent understanding this doesn't feel wrong - you're just explicitly yielding the context. Is there a possible side effect I'm missing?

**Comment by pipermerriam** [comment]
Not that I know of.  Is say proceed with that approach until a better idea
shoes up or it proves problematic.


On Tue, Oct 11, 2016, 2:12 AM rain <notifications@github.com> wrote:

Within Filter.watch add a gevent.sleep(0) to allow the greenlet to spin up.
(This might be a bad idea. It feels a little wrong but it might just work
fine)

In my limited gevent understanding this doesn't feel wrong - you're just
explicitly yielding the context. Is there a possible side effect I'm
missing?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/pipermerriam/web3.py/issues/106#issuecomment-252842523>,
or mute the thread

**Comment by shravan-shandilya** [comment]
Added a PR https://github.com/pipermerriam/web3.py/pull/119

**Comment by pipermerriam** [comment]
Fixed via #119
